dask_cuda/tests/test_local_cuda_cluster.py

```diff
@@ -41,6 +41,7 @@ def get_visible_devices():
                 cluster.scale(1000)
 
 
+@pytest.mark.skipif(utils.get_gpu_count() < 8, reason="Requires 8 GPUs")
```

Comment:
This isn't true. The test may look misleading, but it's actually testing that each process get the `CUDA_VISIBLE_DEVICES` in the correct order, and it can be tested with less GPUs. Let's remove this line.

Reply:
On my laptop with only one GPU and numba v0.45.1 I get:
```python 
CUDA_VISIBLE_DEVICES="2,3,7,8" python tt.py 
Traceback (most recent call last):
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 231, in initialize
    self.cuInit(0)
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 294, in safe_cuda_api_call
    self._check_error(fname, retcode)
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 329, in _check_error
    raise CudaAPIError(retcode, msg)
numba.cuda.cudadrv.driver.CudaAPIError: [100] Call to cuInit results in CUDA_ERROR_NO_DEVICE

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "tt.py", line 5, in <module>
    numba.cuda.get_current_device() 
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/api.py", line 315, in get_current_device
    return current_context().device
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/devices.py", line 213, in get_context
    return _runtime.get_or_create_context(devnum)
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/devices.py", line 139, in get_or_create_context
    return self._get_or_create_context_uncached(devnum)
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/devices.py", line 152, in _get_or_create_context_uncached
    with driver.get_active_context() as ac:
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 387, in __enter__
    driver.cuCtxGetCurrent(byref(hctx))
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 274, in __getattr__
    self.initialize()
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 234, in initialize
    raise CudaSupportError("Error at driver init: \n%s:" % e)
numba.cuda.cudadrv.error.CudaSupportError: Error at driver init: 
[100] Call to cuInit results in CUDA_ERROR_NO_DEVICE:
```
Anything other then `CUDA_VISIBLE_DEVICES="0"` raise a `CUDA_ERROR_NO_DEVICE` exception.

Reply:
Are you sure you don't have some local files that are getting imported by accident and cause the issue? From the log above I see:

```python
Traceback (most recent call last):
  File "tt.py", line 5, in <module>
    numba.cuda.get_current_device() 
```

This file doesn't seem to be part of Numba or dask-cuda. Note also that  this test has been running in CI with a single GPU for months, so I'm inclined to believe that your local environment may have something different or that mobile GPUs behave differently.

Reply:
`tt.py` was just a minimal test.
But you are right, the test actually passes because it ignores the exception. The full tests returns: 
```
(rap) mkristensen@mkristensen:~/repos/dask-cuda$ pytest dask_cuda/tests/test_local_cuda_cluster.py -k test_with_subset_of_cuda_visible_devices -vs
=================================================================================== test session starts ===================================================================================
platform linux -- Python 3.7.3, pytest-4.5.0, py-1.8.0, pluggy-0.11.0 -- /home/mkristensen/anaconda3/envs/rap/bin/python
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/home/mkristensen/repos/dask-cuda/.hypothesis/examples')
rootdir: /home/mkristensen/repos/dask-cuda
plugins: asyncio-0.10.0, hypothesis-4.36.2
collected 3 items / 2 deselected / 1 selected                                                                                                                                             

dask_cuda/tests/test_local_cuda_cluster.py::test_with_subset_of_cuda_visible_devices Unable to start CUDA Context
Traceback (most recent call last):
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 231, in initialize
    self.cuInit(0)
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 294, in safe_cuda_api_call
    self._check_error(fname, retcode)
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 329, in _check_error
    raise CudaAPIError(retcode, msg)
numba.cuda.cudadrv.driver.CudaAPIError: [100] Call to cuInit results in CUDA_ERROR_NO_DEVICE

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mkristensen/repos/dask-cuda/dask_cuda/initialize.py", line 46, in initialize
    numba.cuda.current_context()
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/devices.py", line 213, in get_context
    return _runtime.get_or_create_context(devnum)
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/devices.py", line 139, in get_or_create_context
    return self._get_or_create_context_uncached(devnum)
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/devices.py", line 152, in _get_or_create_context_uncached
    with driver.get_active_context() as ac:
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 387, in __enter__
    driver.cuCtxGetCurrent(byref(hctx))
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 274, in __getattr__
    self.initialize()
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 234, in initialize
    raise CudaSupportError("Error at driver init: \n%s:" % e)
numba.cuda.cudadrv.error.CudaSupportError: Error at driver init: 
[100] Call to cuInit results in CUDA_ERROR_NO_DEVICE:
Unable to start CUDA Context
Traceback (most recent call last):
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 231, in initialize
    self.cuInit(0)
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 294, in safe_cuda_api_call
    self._check_error(fname, retcode)
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 329, in _check_error
    raise CudaAPIError(retcode, msg)
numba.cuda.cudadrv.driver.CudaAPIError: [100] Call to cuInit results in CUDA_ERROR_NO_DEVICE

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mkristensen/repos/dask-cuda/dask_cuda/initialize.py", line 46, in initialize
    numba.cuda.current_context()
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/devices.py", line 213, in get_context
    return _runtime.get_or_create_context(devnum)
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/devices.py", line 139, in get_or_create_context
    return self._get_or_create_context_uncached(devnum)
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/devices.py", line 152, in _get_or_create_context_uncached
    with driver.get_active_context() as ac:
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 387, in __enter__
    driver.cuCtxGetCurrent(byref(hctx))
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 274, in __getattr__
    self.initialize()
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 234, in initialize
    raise CudaSupportError("Error at driver init: \n%s:" % e)
numba.cuda.cudadrv.error.CudaSupportError: Error at driver init: 
[100] Call to cuInit results in CUDA_ERROR_NO_DEVICE:
Unable to start CUDA Context
Traceback (most recent call last):
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 231, in initialize
    self.cuInit(0)
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 294, in safe_cuda_api_call
    self._check_error(fname, retcode)
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 329, in _check_error
    raise CudaAPIError(retcode, msg)
numba.cuda.cudadrv.driver.CudaAPIError: [100] Call to cuInit results in CUDA_ERROR_NO_DEVICE

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mkristensen/repos/dask-cuda/dask_cuda/initialize.py", line 46, in initialize
    numba.cuda.current_context()
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/devices.py", line 213, in get_context
    return _runtime.get_or_create_context(devnum)
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/devices.py", line 139, in get_or_create_context
    return self._get_or_create_context_uncached(devnum)
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/devices.py", line 152, in _get_or_create_context_uncached
    with driver.get_active_context() as ac:
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 387, in __enter__
    driver.cuCtxGetCurrent(byref(hctx))
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 274, in __getattr__
    self.initialize()
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 234, in initialize
    raise CudaSupportError("Error at driver init: \n%s:" % e)
numba.cuda.cudadrv.error.CudaSupportError: Error at driver init: 
[100] Call to cuInit results in CUDA_ERROR_NO_DEVICE:
Unable to start CUDA Context
Traceback (most recent call last):
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 231, in initialize
    self.cuInit(0)
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 294, in safe_cuda_api_call
    self._check_error(fname, retcode)
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 329, in _check_error
    raise CudaAPIError(retcode, msg)
numba.cuda.cudadrv.driver.CudaAPIError: [100] Call to cuInit results in CUDA_ERROR_NO_DEVICE

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mkristensen/repos/dask-cuda/dask_cuda/initialize.py", line 46, in initialize
    numba.cuda.current_context()
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/devices.py", line 213, in get_context
    return _runtime.get_or_create_context(devnum)
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/devices.py", line 139, in get_or_create_context
    return self._get_or_create_context_uncached(devnum)
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/devices.py", line 152, in _get_or_create_context_uncached
    with driver.get_active_context() as ac:
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 387, in __enter__
    driver.cuCtxGetCurrent(byref(hctx))
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 274, in __getattr__
    self.initialize()
  File "/home/mkristensen/anaconda3/envs/rap/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 234, in initialize
    raise CudaSupportError("Error at driver init: \n%s:" % e)
numba.cuda.cudadrv.error.CudaSupportError: Error at driver init: 
[100] Call to cuInit results in CUDA_ERROR_NO_DEVICE:
PASSED
```

Reply:
@pentschev what do you think, should we run the test anyway?
I can write a comment that says the error is expected.

Reply:
I see it now, it only shows when you print the stdout. I think a comment is fine, alternatively, we could add a kwarg to disable passing the `--create-cuda-context` in https://github.com/rapidsai/dask-cuda/blob/branch-0.11/dask_cuda/local_cuda_cluster.py#L111. This would then not cause the context to be created and skip the errors. I'm fine with either, adding a kwarg may be overkill just for this though.